src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+                joinUrl [ baseUrl, "buckets", bucketName, "collections", collectionName, "records", recordId ]
+
+
+client : Config -> Endpoint -> Verb -> Task Error Json.Value
+client config endpoint verb =
+    let
+        request =
+            { verb = verb
+            , url = endpointUrl config endpoint
+            , headers = config.headers
+            , body = Http.empty
+            }
+    in
+        (Http.send Http.defaultSettings request)
+            |> toKintoResponse
+
+
+withHeader : String -> String -> Config -> Config
+withHeader name value ({ headers } as config) =
+    let
+        newHeaders =
+            ( name, value ) :: headers
+    in
+        { config | headers = newHeaders }
+
+
+withAuthHeader : Auth -> Config -> Result Error Config
+withAuthHeader auth config =
+    case auth of
+        NoAuth ->
+            Ok config
+
+        Basic username password ->
+            let
+                hash str =
+                    Base64.encode str |> Result.formatError ConfigError
+
+                addAuthHeader hash =
+                    Ok (withHeader "Authorization" ("Basic " ++ hash) config)
+            in
+                hash (username ++ ":" ++ password)
+                    `Result.andThen` addAuthHeader
+
+        Bearer token ->
+            Ok (config |> withHeader "Authorization" ("Bearer " ++ token))
+
+
+configure : Url -> Auth -> Result Error Config
+configure baseUrl auth =
+    Config baseUrl []
+        |> withAuthHeader auth
+        |> Debug.log "Kinto request"
+
+
+
+-- Dealing with answers from the Kinto server
+
+
+bodyDataDecoder : Json.Decoder Json.Value
+bodyDataDecoder =
+    ("data" := Json.value)
+
+
+extractData : String -> StatusCode -> StatusMsg -> Task Error Json.Value
+extractData body statusCode statusMsg =
+    let
+        responseResult =
+            Json.decodeString bodyDataDecoder body
+    in
+        case responseResult of
+            Ok data ->
+                Task.succeed data
+
+            Err _ ->
+                -- Is it an ErrorRecord json?
+                decodeError body statusCode statusMsg
+
+
+errorDecoder : Json.Decoder ErrorRecord
+errorDecoder =
+    Json.object4 ErrorRecord
+        ("errno" := Json.int)
+        ("message" := Json.string)
+        ("code" := Json.int)
+        ("error" := Json.string)
+
+
+decodeError : String -> StatusCode -> StatusMsg -> Task Error a
+decodeError body statusCode statusMsg =
+    case Json.decodeString errorDecoder body of
+        Ok errorRecord ->
+            Task.fail (KintoError statusCode statusMsg errorRecord)
+
+        Err msg ->
+            Task.fail (ServerError statusCode statusMsg body)
+
+
+toKintoResponse : Task Http.RawError Http.Response -> Task Error Json.Value
+toKintoResponse response =
+    (Task.mapError NetworkError response)
+        `Task.andThen` (handleResponse extractData)
+
+
+handleResponse :
+    (String -> StatusCode -> StatusMsg -> Task Error Json.Value)
```

Comment:
I had to review what I had written and everything else to try to get up to speed on what I was thinking about all this last time.

OK, so staring at this code for a while, I see that you've removed the `extractData` argument to `handleResponse`. The equivalent is the function you've called `extractText`, and now it's embedded in the body of `handleResponse`. I don't have a strong opinion about whether this is better or worse than the existing formulation. You're simplifying code, but at the cost of making it less flexible.

But if we keep this change, and we take a close look at `handleResponse`, we can see that the `extractText` and so forth is really only happening in the "success" branch. Why don't we try to take the call to `extractText` out of `handleResponse` and move it up a level, to `toKintoResponse` ?

If we do that, what remains in `handleResponse`? It's really just a function that examines an HTTP response and tries to pull the body text out of it, and sometimes fails. This is the function I was trying to describe with the name `extractText`. (You can call it whatever you like.) Right now, that logic is tangled up with logic that sometimes calls `extractText` and tries to handle its errors. I think separating it out will provide smaller, more numerous functions, which tends to be a consequence of the functional style.

I see that you also turned `decodeError` into `extractError`, but I think that doesn't actually change anything, right?

`handleResponse` and `extractText` still produce a `Task`, but I still think they should produce a `Result`.

`toKintoResponse` still takes a `Task` as an argument, and I still think that it should just take a `Response` as an argument.